src/Plugins/StateService/StatePlugin.cs

```diff
@@ -102,10 +102,37 @@ void ICommittingHandler.Blockchain_Committing_Handler(NeoSystem system, Block bl
             IReadOnlyList<ApplicationExecuted> applicationExecutedList)
         {
             if (system.Settings.Network != StateServiceSettings.Default.Network) return;
-            StateStore.Singleton.UpdateLocalStateRootSnapshot(block.Index,
+
+            // Rebuild MPT from scratch on node bootstrap if StateService plugin is not up-to-date wrt the current block height.
+            if (block.Index > 0)
+            {
+                using var stateSnapshot = StateStore.Singleton.GetSnapshot();
+                var persistedHeight = block.Index - 1;
+                var stateRoot = stateSnapshot.GetStateRoot(persistedHeight);
+                if (stateRoot is null)
+                {
+                    using var coreSnapshot = system.GetSnapshotCache();
+                    StateStore.Singleton.UpdateLocalStateRootSnapshot(persistedHeight,
+                        coreSnapshot.Find(SeekDirection.Forward)
+                        .Where(si => si.Key.Id != NativeContract.Ledger.Id)
+                        .Select(si => new KeyValuePair<StorageKey, DataCache.Trackable>(si.Key, new DataCache.Trackable(si.Value, TrackState.Added)))
+                        .ToList());
+                    StateStore.Singleton.UpdateLocalStateRoot(persistedHeight);
+                }
+            }
+
+            var root = StateStore.Singleton.UpdateLocalStateRootSnapshot(block.Index,
                 snapshot.GetChangeSet()
                     .Where(p => p.Value.State != TrackState.None && p.Key.Id != NativeContract.Ledger.Id)
                     .ToList());
+            if (system.Settings.IsHardforkEnabled(Hardfork.HF_Faun, block.Index + 1))
+            {
+                var nextH = system.HeaderCache[block.Index + 1];
+                // TODO: the following exception must be critical for the node functioning, need to
+                // refactor this code to stop the node if it happens.
+                if (nextH is not null && root != nextH.PrevStateRoot)
+                    throw new InvalidOperationException($"Stateroot mismatch at {block.Index}: expected {nextH.PrevStateRoot}, got {root}");
```

Comment:
@shargon I need some help for this. This exception must stop the block processing because it means that local stateroot doesn't match the one included into the block by CNs (so either block is invalid or local MPT is invalid, both situations are critical). But I'm not sure how to implement it.

Reply:
I think in this case DoS is exactly what we want. We know the state should be A at height H, if we have state B at the same height something is wrong and node can't be used reliably.